fastlane_bot/tools/optimizer/margpoptimizer.py

```diff
@@ -155,247 +187,288 @@ def optimize(self, sfc=None, result=None, *, params=None):
+                print(f"[margp_optimizer] 1 epsaunit [{epsaunit}] = {p_targettkn_per_epsaunit:,.4f} target [{targettkn}]")
+        crit_is_relative = crit == self.MOCRITR
+        eps_used = eps if crit_is_relative else epsa
+        eps_unit = 1 if crit_is_relative else epsaunit
+        
+        norm = P("norm") or self.MONORML2
+        assert norm in set((self.MONORML1, self.MONORML2, self.MONORMLINF)), f"norm must be MONORML1, MONORML2 or MONORMLINF [{norm}]"
+        normf = lambda x: np.linalg.norm(x, ord=norm)
+        
+        if P("verbose") or P("debug"):
+            print(f"[margp_optimizer] crit={crit} (eps={eps_used}, unit={eps_unit}, norm=L{norm})")
                 
-                # pvec is dict {tkn -> (log) price} for all tokens in p
-                pvec = {tkn: p_ for tkn, p_ in zip(tokens_t, p)}
-                pvec[targettkn] = 1
-                if P("debug") and not quiet:
-                    print(f"pvec={pvec}")
+        ## INNER FUNCTION: CALCULATE THE TARGET FUNCTION
+        def dtknfromp_f(p, *, islog10=True, asdct=False, quiet=False):
+            """
+            calculates the aggregate change in token amounts for a given price vector
+
+            :p:         price vector, where prices use the reference token as quote token
+                        this vector is an np.array, and the token order is the same as in tokens_t
+            :islog10:   if True, p is interpreted as log10(p)
+            :asdct:     if True, the result is returned as dict AND tuple, otherwise as np.array
+            :quiet:     if overrides P("debug") etc, eg for calc of Jacobian
+            :returns:   if asdct is False, a tuple of the same length as tokens_t detailing the
+                        change in token amounts for each token except for the target token (ie the
+                        quantity with target zero; if asdct is True, that same information is
+                        returned as dict, including the target token.
+            """
+            p = np.array(p, dtype=np.float64)
+            if islog10:
+                p = np.exp(p * np.log(10))
+            assert len(p) == len(tokens_t), f"p and tokens_t have different lengths [{p}, {tokens_t}]"
+            if P("debug_dtkn") and not quiet:
+                print(f"\n[dtknfromp_f]\ndtkn=================>>>")
+                print(f"{targettkn:6}", ", ".join(tokens_t))
+                print( "p     ", ", ".join(f"{x:,.2f}" for x in p))
+                print( "1/p   ", ", ".join(f"{1/x:,.2f}" for x in p))
+            
+            # pvec is dict {tkn -> (log) price} for all tokens in p
+            pvec = {tkn: p_ for tkn, p_ in zip(tokens_t, p)}
+            pvec[targettkn] = 1
+            
+            sum_by_tkn = {t: 0 for t in alltokens_s}
+            for pair, (tknb, tknq) in zip(pairs, pairs_t):
+                if get(p, tokens_ix.get(tknq)) > 0:
+                    price = get(p, tokens_ix.get(tknb)) / get(p, tokens_ix.get(tknq))
```

Comment:
You might want to `get(p, tokens_ix.get(tknq))` instead of doing this computation twice.

Reply:
yeah; this always sucks: speed vs garbage collection; more interesting though -- when looking at the code I realized that none of this was needed (it was only produced for logging and logging here is currently commented out) so I removed it...